### PR TITLE
WIP: TST: add tests for Pythran somersd

### PR DESCRIPTION
**Pull request by charlotte12l**
#### Reference issue
https://github.com/scipy/scipy/issues/14544

#### What does this implement/fix?
Take `somersd` as an example, firstly implement all the tests by hand. 

Hi @rgommers, thank you so much for your detailed suggestions! But I'm a little confused about :
>A size-1 array, and (if applicable for the function) a size-0 array
see BUG: RBFInterpolator fails when calling it with a slice of a (1, n) array #14420 for a relevant bug report for non-contiguous inputs and views.

Do you mean size-1 array is like (1, n), and size-0 array is like np.empty(shape=(0, 0)) ? Why would that be a potential problem to Pythran? I think https://github.com/scipy/scipy/issues/14420 is more a problem of views, not a problem of size-1 array..?

For your sketch, 
```
# in class TestMyFunc(PythranFunc)
def pythranfunc(self, x, ):
        "This function is called by the test machinery with various inputs for `x`"
        result = module.funcname(x, ...)  # this is the function call to test, e.g. `stats.somersd`
        expected = # exact result
        assert_allclose(result, expected, rtol=self.rtol, atol=self.atol)
```
How can `pythranfunc` be called by the test machinery with various inputs for `x`? I would expect there is a `_gen_types()` in `PythranFunc`, and in `pythranfunc`, I will call `_gen_types()` to generate different types of `x` and test them?

Also, how many types do you think we should cover...? Currently, there are:
```
        self.ALL_INTEGER = [np.int32, np.int64]
        self.ALL_FLOAT = [np.float32, np.float64]
        self.ALL_COMPLEX = [np.complex64, np.complex128]
```

cc @serge-sans-paille

### CODE REVIEW

**Comment by charlotte12l** [comment]
I made commit https://github.com/scipy/scipy/pull/14559/commits/805777c5e1e2b4bb14362da093db9a54ec96b03f because of `error:cannot collect test class TestSomersD because it has a __init__ constructor`. Maybe `parametrize` is a better way if we can't do `__init__` for test class?

**Comment by rgommers** [comment]
> Do you mean size-1 array is like (1, n), and size-0 array is like np.empty(shape=(0, 0)) ? Why would that be a potential problem to Pythran? I think #14420 is more a problem of views, not a problem of size-1 array..?

Sorry that was incomplete. I think what I had in mind then was a shape `(1,)` array with a large stride. What is special about a size-1 array is that there is only a single element, so the strides can (and sometimes are) be large, or zero. So something like:
```
x = np.array([3])
x.strides = (42,)
```



> I made commit [13f41ed](https://github.com/scipy/scipy/commit/13f41edd87a8029dcdd8c15fc9d96fbfc06dda7d) because of `error:cannot collect test class TestSomersD because it has a __init__ constructor`. Maybe `parametrize` is a better way if we can't do `__init__` for test class?

Ah yes, there is no `__init__` because pytest does something special. There is a `setup` method though that is always called first, so that can be used instead.

The problem with doing `for dtype in [np.int32, np.int64, np.float32, np.float64]:` is that it gets repeated over and over. You can use `parametrize` instead of a for-loop if you prefer, that's a bit of a judgement call (it will create multiple separate test cases instead of one, not sure that that's better here). But either way you want to refer to a `self.dtypes` to not repeat yourself.

**Comment by rgommers** [comment]
> how many types do you think we should cover

It would maybe be excessive to call all of them, but let's at least do a few that Pythran does not have the signature for, like `uint8`, `int16` in addition to `int32`, `int64`.

**Comment by rgommers** [comment]
> How can `pythranfunc` be called by the test machinery with various inputs for `x`? I would expect there is a `_gen_types()` in `PythranFunc`, and in `pythranfunc`, I will call `_gen_types()` to generate different types of `x` and test them?

I was thinking it's a special name, so `PythranFunc` will already know how to call it and does so with the various inputs under test. So `PythranFunc` willl have something like:
```python
    def test_all_dtypes(self):
        _x = np.arange(10)
        for dt in self.dtypes:
            x = _x.astype(dt)
            self.pythranfunc(x)

    def test_views(self):
        ... (etc)
```

> I will call `_gen_types()` to generate different types of `x` and test them?

You can try this too. The goal is to be to write tests for new functions that don't have unnecessary boilerplate but are still easy enough to understand.

**Comment by charlotte12l** [comment]
> I was thinking it's a special name, so `PythranFunc` will already know how to call it and does so with the various inputs under test. So `PythranFunc` willl have something like:
> 
> ```python
>     def test_all_dtypes(self):
>         _x = np.arange(10)
>         for dt in self.dtypes:
>             x = _x.astype(dt)
>             self.pythranfunc(x)
> 
>     def test_views(self):
>         ... (etc)
> ```
> 

Thanks for your detailed reply! But I think it will not work well if we define `x` in `PythranFunc`(  `_x = np.arange(10)`), because for different functions, the input shape may vary, e.g. 1d/2d/3d. Moreover, `self.pythranfunc(x)` was defined in the Child class and it is hard to check whether the result meets the expectation, because the child class can't know the value of `x` which is defined in the parent class `PythranFunc` .  Unless we do something like this? (not sure if it works):

```
class PythranFunc:
    '''
    Inherit from this class to generate a bunch of relevant test cases
    '''
    def __init__(self):
        self.ALL_INTEGER = [np.int8, np.int16, np.int32, np.int64]
        self.ALL_FLOAT = [np.float32, np.float64]
        self.ALL_COMPLEX = [np.complex64, np.complex128]
        self.dtypes = [] # updated in child class
        self.arguments = [] # updated in child class
        self.index = []# updated in child class
        self.expected = None # updated in child class
    
    def test_all_dtypes(self):
        for dtype in self.dtypes:
            self.arguments[self.index] = self.arguments[self.index].astype(dtype)
        res = self.pythranfunc(self.arguments)
        assert_allclose(res.statistic, self.expected[0], atol=1e-15)
        assert_allclose(res.pvalue, self.expected[1], atol=1e-15)

```

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ jobs:
 
   test_nightly:
     name: Nightly CPython
-    if: "github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[ci skip]') && !contains(github.event.head_commit.message, '[skip ci]') && !contains(github.event.head_commit.message, '[skip github]') && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/')"
+    if: "github.repository == 'charlotte12l/scipy' && !contains(github.event.head_commit.message, '[ci skip]') && !contains(github.event.head_commit.message, '[skip ci]') && !contains(github.event.head_commit.message, '[skip github]') && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/')"
```
Comment (rgommers) [inline .github/workflows/linux.yml:56]: Can you undo these changes?

Comment (charlotte12l) [reply .github/workflows/linux.yml]: Sorry, I forgot I've changed linux.yml on my master when I rebase on my master...

File: scipy/_lib/_testutils.py [hunk 2]
```diff
@@ -71,6 +73,73 @@ def __call__(self, label="fast", verbose=1, extra_argv=None, doctests=False,
         return (code == 0)
 
 
+class _TestPythranFunc:
+    '''
+    Inherit from this class to generate a bunch of relevant test cases
+    '''
+    ALL_INTEGER = [np.int8, np.int16, np.int32, np.int64]
```
Comment (rgommers) [inline scipy/_lib/_testutils.py:80]: Last week we discussed whether it is useful/needed to add `intc`/`intp` here, and I thought the outcome was yes.

File: scipy/_lib/_testutils.py [hunk 3]
```diff
@@ -71,6 +73,73 @@ def __call__(self, label="fast", verbose=1, extra_argv=None, doctests=False,
...
+        for dtype in self.dtypes:
+            # keep a copy of self.arguments to avoid in-place modify
+            test_array = self.arguments.copy()
+            for idx in self.index:
+                test_array[idx] = test_array[idx].astype(dtype)
+            self.pythranfunc(*test_array)
+
+    def test_views(self):
+        # keep a copy of self.arguments to avoid in-place modify
+        test_array = self.arguments.copy()
```
Comment (rgommers) [inline scipy/_lib/_testutils.py:114]: These two lines are now present in every test case, however I don't see where the in-place modification would occur. Is this really necessary? No well-designed function should modify any of its input arguments.

File: scipy/stats/tests/test_hypotests.py [hunk 4]
```diff
@@ -614,7 +614,18 @@ def test_gh_2118(self, x, y, alternative, expected):
         assert_allclose(res, expected, rtol=1e-12)
 
 
-class TestSomersD:
+class TestSomersD(_TestPythranFunc):
+    def setup_method(self):
+        self.dtypes = self.ALL_INTEGER + self.ALL_FLOAT
+        self.arguments = [np.arange(10), np.arange(10)]
+        self.index = [0, 1]
```
Comment (rgommers) [inline scipy/stats/tests/test_hypotests.py:621]: Is `self.index` always (`range(len(self.arguments))`)? If so, it does not have to be specified separately.

**Review by rgommers — COMMENTED** [review]
This is starting to look good. I asked some questions to hopefully make the base class a little more concise. It already tests what we want to test though, and `TestSomersd` looks nice and simple.

File: scipy/stats/tests/test_hypotests.py (hunk 4 above)
Comment (charlotte12l) [reply scipy/stats/tests/test_hypotests.py:621]: No, sometimes the function's arguments are not of interest, like `nonetype`, `str`, and we should not test the (float/int) dypes for such arguments. Also, it came to me that different arguments may have different dtypes to test?

File: scipy/_lib/_testutils.py (hunk 3 above)
Comment (charlotte12l) [reply scipy/_lib/_testutils.py:114]: Ah.. I thought if I didn't do `test_array = self.arguments.copy()`, then `self. arguments` will be modified `self. arguments[idx] = self. arguments[idx][::-1][::-1]` ?  and I think the best way is to copy the array first, and then modified the some of the elements?

File: scipy/stats/tests/test_hypotests.py (hunk 4 above)
Comment (rgommers) [reply scipy/stats/tests/test_hypotests.py:621]: If an argument is not of interest, maybe it should not be given in this list? That can always be avoided by using `functools.partial`.

The case that I'm not sure needs to be handled is:
```
def somefunc(x0, str_param, x1):
```
That pattern is so unusual that it can be ignored I think.

> Also, it came to me that different arguments may have different dtypes to test?

That does seem more likely to happen. So maybe all this can be a single dictionary? Something like:
```
self.arguments = {0: (np.arange(10), (self.ALL_INTEGER + SELF.ALL_FLOAT)),
                  1: ...}
```

File: scipy/_lib/_testutils.py (hunk 3 above)
Comment (rgommers) [reply scipy/_lib/_testutils.py:114]: Okay yes I see it now, sorry for the noise. One other suggestion, to avoid repeating the same comment everywhere: use a separate method, `args = self._get_args()`. With implementation:
```
def _get_args(self):
    # returns a copy of self.arguments because some of the test functions in this
    # class perform an in-place modification of the argument list before passing those
    # arguments to the function under test.
    return self.arguments.copy()

File: scipy/stats/tests/test_hypotests.py (hunk 4 above)
Comment (charlotte12l) [reply scipy/stats/tests/test_hypotests.py:621]: > If an argument is not of interest, maybe it should not be given in this list? 

Yes! So I was using `self.index` to store only the index of the argument of interest(then the length is not always (range(len(self.arguments)))), and it can also handle the case `def somefunc(x0, str_param, x1):`.

> That does seem more likely to happen. So maybe all this can be a single dictionary?

Make sense! A single dictionary is absolutely better than using several lists, I'll use that :)

Comment (charlotte12l) [reply scipy/stats/tests/test_hypotests.py:621]: Ah, no , `self.arguments` are supposed to store all the arguments needed for that function(including that not of interests), and if we used dictionary, we can't know those arguments because the key(0, 1, ...) is only those arguments of interests. Moreover, we can do `self.pythranfunc(*self.arguments)` if it is a list, otherwise we need to create a new list every time? (This is also okay I think)

Comment (charlotte12l) [reply scipy/stats/tests/test_hypotests.py]: Now I switched to use a single dictionary, and I made the following changes(not sure whether it is reasonable):
- To test different dtypes for different input arrays, e.g. `def func(a,b)`, `a` have 5 dtypes to test, `b` have 3 dtypes to test.
 I don't think we should test all the `5*3` combinations of `a` dtypes and `b` dtypes because `a` and `b` would not influence each other. As long as `int` `a` & `int` `b` pass tests, `float` `a` & `float` `b` works, then `int` `a` & `float` `b` will definitely work too! (Unless there limits `a` and `b` must be the same type, but that's not what Pythran could do). Therefore, in my implementation, we only need to test `5` times rather than `5*3`. We iterate on `for idx in range(5)` types and for `b` which only has 3 types to test, it will keep using the last dtype for `idx>3`. See `get_max_dtype_list_length`, `get_dtype`, `test_all_dtypes` for details.
- To test keywords. e.g. Test Somersd `def somersd(x, y=None)`.
In my understanding, testing keywords means testing whether the results of specifying the keyword arguments(`somersd(x, None)`) vs not-specifying the keyword arguments (`somersd(x)`) are the same.
Therefore, I found it difficult to make `test_keywords` similar to `test_dtypes`/`test_views`/`test_stride`. For `def somersd(x, y=None)`, `y` can also be an array that requires testing dtypes or views, and if `y` is not None, `x` and `y` must be 1d array. If `y` is None, `x` must be a 2d array. In this way, when we `test_dtypes`/`test_views`, the input args are two 1d array, when we `test_keywords`, the input arg is one 2d array. It would be confusing if we defined so many different `self.arguments`/ `self.nonoptional_arguments` in the child class. Therefore, I choose to move `test_keywords` to the child class rather than remain in the parent class, but we provide util functions in the parent class like `self.get_optional_args` for `test_keywords` to use.

**Comment by charlotte12l** [comment]
There are some failures in Mypy but seems it is unrelated.

**Review by tylerjereddy — APPROVED** [review]
Well, Ralf was reasonably positive about the state of the PR and it just adds new testing infrastructure without modifying any old tests so I don't mind putting it in.

I did find it a bit confusing to skim over--for example, is the infrastructure guaranteed to work gracefully for cases where Pythran is deactivated and we use the Cython implementation. Maybe I missed that explanation somewhere.
